Kajabi/sage-archived
# Stylelint configuration

## Description
Sage CSS linting configuration. Follows patterns drawn from UXD style survey.

VSCode:
![linting-vscode](https://user-images.githubusercontent.com/816579/76019069-03eb8380-5ed6-11ea-8831-b20b42a8ea0b.png)

Atom:
![linting-atom](https://user-images.githubusercontent.com/816579/76018925-c5ee5f80-5ed5-11ea-93c8-84117cd32427.png)

Rules/guidelines
**default violations give a warning, rules marked with a * are enforced and throw an error**

- [x] newline before/after brackets in declaration blocks*
- [x] newline per declaration
- [x] no empty blocks *
- [x] space between properties and values
- [x] enforce semicolon use when ending declarations
- [x] favor double-quotes (`"`) over single-quotes (`'`) *
- [x] enforce leading zeroes in decimal values *
- [x] lowercase property names *
- [x] lowercase values *
- [x] single space between values and properties *
- [x] pseudo-elements require double colon (`::`)
- [x] font-family names should be surrounded by quotes(`"`) *
- [x] "named" colors (ex. `black`, `skyblue`) are only allowed within functions
- [x] no empty comments (`/* */`) *
- [x] 3 level max nesting depth
- [x] whitespace between comment opening/closing marks and contents 
- [x] duplicate properties are _not_ allowed unless consecutive (ex. CSS custom property fallbacks) *
- [x] limit of 2 empty lines in declaration blocks
- [x] animations that aren't defined in `@keyframes` disallowed *
- [x] single space required between multiple comma-separated `@media` declarations *
- [x] disallow use of `!important` (can be overridden with `/* stylelint-ignore */`) in special cases

## Test notes

### Steps for testing
An example with comments for rules/guidelines can be found [here](https://gist.github.com/monkeypox8/88424b7baf6a2e9de86cd766d4f314c1).

Linting can be performed from the command line, but is much more convenient via your [editor of choice](https://stylelint.io/user-guide/integrations/editor).

CLI steps:
1. Run `bundle` and `npm i` from the parent `sage` directory to install dependencies
2. `npx stylelint <stylesheet file path>` to lint files. ex: `npx stylelint "app/assets/**/*.scss"` will check and lint the entire assets directory (warning: it's a long list)

Code editor steps:
1. Run `bundle` and `npm i` from the parent `sage` directory to install dependencies
2. Download, install, and enable the appropriate integration for your editor from [stylelint](https://stylelint.io/user-guide/integrations/editor)
3. Open a `.scss` file in Sage
4. ???
5. [Profit](https://www.youtube.com/watch?v=tO5sxLapAts)

## Related issues
-  Resolves #27
